GPU role (first part)

scimerman opened the pull request:
This is the first part of the role: driver installation on GPU compute nodes.
Did not have time to extensively test it, but so far the GPU's worked. Tested on a one GPU and multiple GPU machine, plus on a few non-gpu machines.
First part of Readme provided.
The software (libraries and compilers) has not yet been implemented.

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,67 @@
+      - wget
+  become: true
+
+# ansible_kernel variable is not working, as after reboot, still holds old kernel
```
Can this be handled with a task to update a specific "gathered fact". E.g. something like this:
https://stackoverflow.com/questions/35185635/can-ansible-fetch-updated-facts-in-middle-of-a-playbook

scimerman replied:
Is running `ansible.builtin.setup` step any better than running `uname -r`?
For uname I can at least trust that it will always work :smile:. Of course I can change it, if you think this is better in any way.

scimerman replied:
Updated to setups :heavy_check_mark:

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,67 @@
+  register: driver_downloaded
+  become: true
+
+- name: Install driver from .run file
```
This is not idempotent and will always re-install the driver also when the same version is already installed...

pneerincx replied:
Ok, point taken, but in that case you can drop the  "Check if driver downloaded" task. If the download failed with the get_url task, the play will already stop and never reach  "Check if driver downloaded".

scimerman replied:
updated

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,67 @@
+  ansible.builtin.command: '/root/{{ gpu_runfile }} --silent --driver'
+  register: install_result
+  failed_when: install_result.rc != 0
+  when: driver_downloaded.stat.exists
```
Is this "when" necessary? If the driver was not successfully downloaded (either now or during a previous play), this task should fail. Instead it will now get skipped and nothing will be installed and there will not be an error either.

scimerman replied:
Ansible-lint expects the steps with `ansible.builtin.command` to have `when` defined. I guess this is the reasoning on why I add it here...
If the driver download would fail, then this (as the step before installing) would fail first.
I don't know if it make sense to change it into `when: true`?

pneerincx replied:
Ansible-lint wants you to make sure the task is idempotent, which requires "something" to determine if the command should be executed or skipped, because it has already been used successfully before. Adding `when` is one way to make it idempotent, but condition you added for the `when` does not make sense. If this `gpu.yml` is only included when a GPU was detected and drivers are missing, then that ensures idempotency and you can just use `when: true` here.

scimerman replied:
fixed

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,67 @@
+  notify: reboot_server
+  become: true
+
+- name: Remove installation file
```
Why remove the installer? That means it will have to be re-downloaded again each time the role is executed.

scimerman replied:
Well, alternative is to use 3.3GB on every GPU machine, even after the installation was successful. If the installation would be small, then it can be easily downloaded- if large, then it needs to be downloaded.
When it comes to this step, the driver should be already installed, as all went fine. And so when this role is finalized, there should be no more reason to keep the file. It either works, or it doesn't and on-machine debugging needs to be done - manually.

scimerman replied:
What I do realize, is that I should have at the end the same (as the one at the beginning) checkup with `nvidia-smi` and check if all devices are listed there.

pneerincx replied:
Ok, 3.3 GB is quite a large file for the root partition. In that case I understand why it is removed again. Would be nice though to be able to _cache_ this somehow for a cluster just like we do with RPMs using Pulp. Would be bad if a compute node needs to be re-installed and the version of a driver is no longer available for download for whatever reason.

pneerincx replied:
As this GPU related stuff is work in progress anyway, I 'm happy to accept this as is with deleting the driver download, but it would be nice to see if we can use the `file` plugin caching of these driver files with Pulp in a future PR. We already install the `file` plugin in addition to the `rpm` plugin, but have not used it yet.

pneerincx reviewed (changes requested):
See inline questions/comments.

scimerman commented:
I tried to reduce the overhead, so now what is left, are actually needed parts.
There was an attempt to install the driver from smaller size file, but unfortunately that installation is without `nvidia-smi` and automatic driver initialization tools. So I ended up back again with 3GB large cuda .run installation file, that works.
The automatic detection of GPU devices was failing in number of occasions (if interested, check readme - Solved issues). 
To solve the problem, I added a variable called `gpu_count` - it is set for each GPU node, and it defines expected number of running GPUs.

Role tested on gpu and non-gpu instances.

pneerincx commented on roles/gpu/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## Role outline
+
+- it expects the gpu_count to be defined per invididual machine
```
```suggestion
- it expects the `gpu_count` variable to be defined per invididual machine.
```

pneerincx commented on roles/gpu/README.md:
```diff
@@ -0,0 +1,68 @@
+## Role outline
+
+- it expects the gpu_count to be defined per invididual machine
+- attempts to gather the GPU device status by running `nvidia-smi` command
```
```suggestion
- runs `nvidia-smi` command to gather the GPU device status.
```

pneerincx commented on roles/gpu/README.md:
```diff
@@ -0,0 +1,68 @@
+   - `nvidia-smi` command is not available (cuda driver was not installed)
+   - `nvidia-smi` reports different number of GPU devices than expected from `gpu_count`
+- yum install on machine packages that is needed for driver install and compile
+- downloads the cuda .run driver file from nvidia website (version defined in defualts)
```
```suggestion
- downloads the cuda `*.run` driver file from Nvidia website (version defined in defaults).
```

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,99 @@
+- name: Install kernel developement package matching running kernel version
+  ansible.builtin.yum:
+    name: 'kernel-devel-{{ ansible_kernel }}'
+  register: yum_result
+  failed_when: yum_result.rc != 0
+  when: true
```
```suggestion
```

pneerincx commented on roles/gpu/tasks/main.yml:
```diff
@@ -0,0 +1,24 @@
+- name: Reboot system if needed
+  ansible.builtin.meta: flush_handlers
+
+- name: Check how many NVidia devices is up and running (might take some time)
```
```suggestion
- name: Check how many NVidia devices are up and running (might take some time)
```

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,99 @@
+    dest: '/root/{{ gpu_runfile }}'
+    mode: '0700'
+  become: true
+
```
```suggestion
  when: (smi_gpu_count.stdout | default([]) | lower | regex_findall('nvidia') | length != gpu_count) or
             (smi_driver_version.stdout  | default([]) != gpu_driver_version)

```

scimerman replied:
again, this is not needed if we avoid move into `gpu.yml`

pneerincx commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,99 @@
+  ansible.builtin.command: '/root/{{ gpu_runfile }} --silent --driver'
+  register: install_result
+  failed_when: install_result.rc != 0
+  when: true
```
```suggestion
  when: (smi_gpu_count.stdout | default([]) | lower | regex_findall('nvidia') | length != gpu_count) or
             (smi_driver_version.stdout  | default([]) != gpu_driver_version)
```

scimerman replied:
same, can be avoided if move into `gpu.yml` is omitted

pneerincx commented on roles/gpu/tasks/main.yml:
```diff
@@ -0,0 +1,24 @@
+---
+- name: Check if system needs to be restarted
+  ansible.builtin.command: '/bin/needs-restarting -r'
+  register: needs_restarting
+  failed_when: 'needs_restarting.rc > 1'
+  changed_when: 'needs_restarting.rc == 1'
+  become: true
+  notify: reboot_server
+
+- name: Reboot system if needed
+  ansible.builtin.meta: flush_handlers
+
+- name: Check how many NVidia devices is up and running (might take some time)
+  ansible.builtin.command: 'nvidia-smi -L'
+  register: smi
+  changed_when: false
+  failed_when: false
+  become: false    # running nvidia-smi as root stops the service
+
+- name: Install GPU driver if not all GPU devices are present and working
+  ansible.builtin.include_tasks: gpu.yml
+  when: ( gpu_count is defined ) and
+        ( smi.stdout|default([])|lower|regex_findall('nvidia')|length != gpu_count )
```
```suggestion
- name: Install GPU driver if one or more GPUs were specified for this machine.
  ansible.builtin.include_tasks: gpu.yml
  when: gpu_count is defined and gpu_count | length >= 1
```

scimerman replied:
I do not like this being moved into `gpu.yml`. 
`gpu_count` on it own does not say define anything about status of the system.
`nvidia-smi` is the only way to reliable detect status of gpu devices.
Therefore we need to run the `nvidia-smi`, and based on this result decide if we need to (re)install gpu driver.

If `nvidia-smi` is executed on non-gpu machine - fine. I will detect (together with `gpu_count` that this machine needs to be skipped).

scimerman replied:
We could put into `gpu.yml` the "Check if system needs to be restarted".
But then again, it is run only if it needed. So I see no harm running it - perhaps quite opposite.

pneerincx replied:
> gpu_count on it own does not say define anything about status of the system.
> nvidia-smi is the only way to reliable detect status of gpu devices.

This contracts what was written in the README.md:
> `gpu_count` is needed to install the driver, since any other `automatic` detection is
> failing sooner or later.

This includes `nvidia-smi` reporting wrong values. Either `nvidia-smi` or some other tool automagically detects whether GPUs are present and the driver needs to be installed or we define the expected state in a variable like `gpu_count` and rely on that.

Even if `nvidia-smi` would report the correct number of GPUs, we should not skip `gpu.yml`. The driver may be installed, but there are various other tasks that need to re-run to ensure idempotency:
* Check if correct version of the driver was installed.
* Create nvidia group and user with correct UID and GID.
* Install blacklist-nouveau.conf
* etc.

scimerman replied:
> This contracts what was written in the README.md:
And all that is correct. No changes of readme needed. Perhaps a word or two added, if it is not clear.
> This includes nvidia-smi reporting wrong values. Either nvidia-smi or some other tool automagically detects whether GPUs are present and the driver needs to be installed or we define the expected state in a variable like gpu_count and rely on that.

No. This are two different thing. 
  - `gpu_count` is explaining the *count* of working gpu's expected. And this cannot be automatically detected - reliably
  - (if all is working as it should) `nvidia-smi` detects how many working gpu's there are

=========================================================
Long story short; main question is:
- is it ok to skip the configuration part when all GPU's are up and running?

I advocate for yes.

=========================================================
Long story long:
> Even if nvidia-smi would report the correct number of GPUs, we should not skip gpu.yml.

I think if everything works, we should skip the whole thing. That was the entire purpose of detecting with nvidia-smi. If it works, then there is nothing to be configured: the service is up and running, blacklisting of non-nvidia driver was done, nvidia user was created ...

> Check if correct version of the driver was installed.

I don't see how the driver version will change on it's own. If all working, why do we need to check driver version?

> Create nvidia group and user with correct UID and GID.

Without nvidia user and nvidia group, nothing works -> gpu.yml gets triggered

scimerman commented on roles/gpu/tasks/gpu.yml:
```diff
@@ -0,0 +1,99 @@
+      - bind-utils
+      - wget
+  become: true
+
```
see the other comment about moving into `gpu.yml`

pneerincx commented on roles/gpu/tasks/main.yml:
```diff
@@ -0,0 +1,25 @@
+        gpu_driver_version not in modinfo.stdout|default("")|regex_search("version:.*"))
+
+- name: Configure user and services
+  ansible.builtin.include_tasks: user_services.yml
```
`user_services.yml` does not exist in this PR and `gpu.yml` does exist but is no longer used anywhere as far as I can tell. Looks like one or more commits are missing...

pneerincx reviewed (changes requested):
See inline comment.

scimerman commented:
@pneerincx I have addressed the issues you have raised, and deploy several times the playbook on GPUs. Playbook fails on the missing storage, which it seems to be 'coming soon ...' for too long now. I would recommend that we merge what we have done so far.
Another PR is comming for the ansible-pipelines - added easybuild PyTorch and TensorFlow.
